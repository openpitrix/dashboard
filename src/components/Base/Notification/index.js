import Item from './item';
import Notification from './manager';

Notification.Item = Item;

export default Notification;
