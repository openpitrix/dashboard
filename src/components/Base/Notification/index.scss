
$bg-info: lighten(blue, 30%);
$bg-success: green;
$bg-warning: yellow;
$bg-error: red;

.notification {
  position: absolute;
  top: 20px;
  right: 20px;
  width:220px;
  min-height:60px;
  padding: 3px 10px;
  font-size:14px;
  line-height:20px;
  color: #fff;
  box-sizing: border-box;
  border-radius: 6px;
  //opacity: .8;
  animation-duration: 2s;
  animation-name: slidein;

  .title {
    display: inline-block;
    font-size: 1.2em;
    margin-bottom: 10px;
  }

  .message {
    text-indent:1em;
    padding-bottom:10px;
  }

  &-info {
    background: $bg-info;
  }
  &-success {
    background: $bg-success;
  }
  &-warning {
    background: $bg-warning;
  }
  &-error {
    background: $bg-error;
  }
}