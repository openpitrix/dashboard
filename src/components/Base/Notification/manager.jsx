import React from 'react';
import PropType from 'prop-types';
import classnames from 'classnames';
import { observer, inject } from 'mobx-react';

import NotificationItem from './item';

import styles from './index.scss';

@inject('rootStore')
@observer
export default class NotificationManager extends React.Component {
  static propTypes = {
    className: PropType.string,
    itemClass: PropType.string
  };

  componentWillUnmount() {
    this.props.rootStore.notifications = [];
  }

  render() {
    const { className, itemClass, rootStore } = this.props;
    const { notifications, detachNotify } = rootStore;

    return (
      <div className={classnames(styles.wrapper, className)}>
        {notifications.map((notification, idx) => (
          <NotificationItem
            className={itemClass}
            {...notification}
            key={idx}
            detach={detachNotify}
          />
        ))}
      </div>
    );
  }
}
