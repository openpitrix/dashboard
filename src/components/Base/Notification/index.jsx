import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { Icon } from 'components/Base';
import styles from './index.scss';

export default class Notification extends React.Component {
  static propTypes = {
    type: PropTypes.oneOf(['info', 'success', 'warning', 'error']),
    title: PropTypes.string,
    message: PropTypes.node,
    timeOut: PropTypes.number,
    onClick: PropTypes.func,
    onHide: PropTypes.func,
    onClosed: PropTypes.func
  };

  static defaultProps = {
    type: 'error',
    title: 'Notification',
    message: null,
    timeOut: 2000,
    onClick: () => {},
    onHide: () => {},
    onClosed: () => {}
  };

  timer = null;

  // based on font-awesome icons
  iconMap = {
    error: 'error',
    success: 'information',
    info: 'information',
    warning: 'exclamation'
  };
  colorMap = {
    error: '#cf3939',
    success: '#15934e',
    info: '#1e6eeb',
    warning: '#ef762b'
  };

  componentDidMount() {
    const { timeOut, onHide } = this.props;
    if (timeOut) {
      this.timer = setTimeout(onHide, timeOut);
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
    this.timer = null;
    this.props.onClosed();
  }

  handleClick = e => {};

  render() {
    const { type, message } = this.props;
    const className = classnames(styles.notification, styles[`notification-${type}`]);
    const colorStyles = {
      primary: '#fff',
      secondary: this.colorMap[type]
    };
    return (
      <div className={className} onClick={this.handleClick} ref={c => (this.target = c)}>
        <Icon name={this.iconMap[type]} size={18} color={colorStyles} />
        {message}
      </div>
    );
  }
}
