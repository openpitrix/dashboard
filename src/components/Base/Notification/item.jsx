import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { noop } from 'lodash';

import { Icon } from 'components/Base';

import styles from './index.scss';

export default class NotificationItem extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    detach: PropTypes.func,
    message: PropTypes.node,
    onClick: PropTypes.func,
    onClosed: PropTypes.func,
    onHide: PropTypes.func,
    timeOut: PropTypes.number,
    title: PropTypes.string,
    ts: PropTypes.number,
    type: PropTypes.oneOf(['info', 'success', 'warning', 'error'])
  };

  static defaultProps = {
    type: 'info',
    title: 'Notification',
    message: null,
    timeOut: 3000,
    onClick: noop,
    onHide: null,
    onClosed: noop,
    detach: noop,
    ts: Date.now()
  };

  timer = null;

  state = {
    hidden: false
  };

  // based on font-awesome icons
  iconMap = {
    error: 'error',
    success: 'information',
    info: 'information',
    warning: 'exclamation'
  };

  colorMap = {
    error: '#cf3939',
    success: '#15934e',
    info: '#1e6eeb',
    warning: '#ef762b'
  };

  hideNotify = () => {
    const { ts, detach, onHide } = this.props;

    this.setState({ hidden: true }, () => {
      // detach notification from mobx store
      detach(ts);

      onHide && onHide();
    });
  };

  componentDidMount() {
    const { timeOut } = this.props;
    if (timeOut) {
      this.timer = setTimeout(this.hideNotify, timeOut);
    }
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
    this.timer = null;
    this.props.onClosed();
  }

  handleClick = e => {
    // this.hideNotify();
  };

  render() {
    const { type, message, className } = this.props;
    const { hidden } = this.state;

    const colorStyles = {
      primary: '#fff',
      secondary: this.colorMap[type]
    };

    if (hidden) {
      return null;
    }

    return (
      <div
        className={classnames(
          styles.notification,
          styles[`notification-${type}`],
          className
        )}
        onClick={this.handleClick}
        ref={c => {
          this.target = c;
        }}
      >
        <Icon name={this.iconMap[type]} size={18} color={colorStyles} />
        {message}
      </div>
    );
  }
}
