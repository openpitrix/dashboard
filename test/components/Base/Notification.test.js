import { Provider } from 'mobx-react';
import { observable } from 'mobx';

import Notification from 'components/Base/Notification';

describe('Base/Notification', () => {
  const store = observable.object({
    notifications: [{ ts: 1, message: 'note1' }, { ts: 2, message: 'note2' }]
  });
  it('basic render', () => {
    const wrapper = render(
      <Provider rootStore={store}>
        <Notification />
      </Provider>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
