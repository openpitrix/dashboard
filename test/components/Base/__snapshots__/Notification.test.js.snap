// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Notification basic render 1`] = `
<div
  class="wrapper"
>
  <div>
    <div
      class="notification notification-info"
    >
      <span
        class="icon icon-information"
        style="width:18px;height:18px"
      >
        <svg
          class="qicon qicon-information qicon-light"
          style="--primary-color:#fff;--secondary-color:#1e6eeb"
        >
          <use
            href="#qui-information"
          />
        </svg>
      </span>
      note1
    </div>
    <div
      class="notification notification-info"
    >
      <span
        class="icon icon-information"
        style="width:18px;height:18px"
      >
        <svg
          class="qicon qicon-information qicon-light"
          style="--primary-color:#fff;--secondary-color:#1e6eeb"
        >
          <use
            href="#qui-information"
          />
        </svg>
      </span>
      note2
    </div>
  </div>
</div>
`;
